src/features/Folding.ts

```diff
@@ -0,0 +1,415 @@
+                                          MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching Braces   #region -> #endregion
+        this.matchScopeElements(this.extractRegionScopeElements(tokens, document),
+                                "custom.start.region",
+                                "custom.end.region",
+                                MatchType.Region, document)
+            .forEach( (x) => { matchedTokens.push(x); });
+
+        // Find blocks of line comments   # comment1\n# comment2\n...
+        this.matchBlockCommentScopeElements(tokens, document).forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching function definitions   <# -> #>
+        this.matchScopeElements(tokens,
+                                "punctuation.definition.comment.block.begin.powershell",
+                                "punctuation.definition.comment.block.end.powershell",
+                                MatchType.Comment, document)
+            .forEach( (x) => { matchedTokens.push(x); });
+
+        return matchedTokens;
+    }
+
+    // Search all of the loaded extenions for the PowerShell grammar file
+    private powerShellGrammarPath(): string {
+        try {
+            const psGrammars =
+                vscode.extensions.all
+                    .filter((x) => x.packageJSON && x.packageJSON.contributes && x.packageJSON.contributes.grammars)
+                    .reduce((a: IExtensionGrammar[], b) =>
+                        [...a, ...(b.packageJSON as IExtensionPackage).contributes.grammars
+                            .map((x) => Object.assign({ extensionPath: b.extensionPath }, x))], [])
+                    .filter((x) => x.language === "powershell");
+            if (psGrammars.length === 0) { return ""; }
+            return path.join(psGrammars[0].extensionPath, psGrammars[0].path);
+        } catch (err) { return ""; }
+    }
+
+    // Returns the PowerShell grammar parser
+    private grammar() {
+        if (this.powershellGrammar !== undefined) { return this.powershellGrammar; }
+        const registry = new tm.Registry();
+        this.powershellGrammar = registry.loadGrammarFromPathSync(this.powerShellGrammarPath());
+        return this.powershellGrammar;
+    }
+
+}
+
+export class FoldingFeature implements IFeature {
+    private foldingProvider: FoldingProvider;
+
+    constructor(documentSelector: DocumentSelector) {
+        // Hide the syntax folding behind a feature flag
+        const settings = Settings.load();
+        if (settings.developer && settings.developer.featureFlags) {
+            if (settings.developer.featureFlags.indexOf("syntax-folding") === -1) { return; }
+        }
+
+        this.foldingProvider = new FoldingProvider();
+        vscode.languages.registerFoldingRangeProvider(
+            documentSelector,
+            this.foldingProvider);
+    }
+
+    // tslint:disable-next-line:no-empty
```

Comment:
It looks like these tslint pragmas might be necessary, but I would like to get rid of them if possible. I see there are [recommendations to return `undefined` or `void`](https://stackoverflow.com/questions/31605781/why-do-tslint-and-jslint-report-empty-blocks), but I'm not sold on that either. Thoughts?